subsys/mgmt/mcumgr/grp/fs_mgmt/Kconfig

```diff
@@ -161,4 +161,54 @@ config MCUMGR_GRP_FS_FILE_ACCESS_HOOK
 	  Note that this will be called multiple times for each file read and
 	  write due to mcumgr's method of stateless operation.
 
+config MCUMGR_GRP_FS_FILE_SEMAPHORE_TAKE_TIME
+	int "File handle semaphore take time (ms)"
+	default 100
+	help
+	  Maximum time (in ms) to acquire the file handle semaphore when a file
+	  upload/download command is used. If unable to acquire the semaphore,
+	  then MGMT_ERR_EBUSY will be returned.
+
+	  Can specify 0 to not wait for the lock and return instantly if it
+	  cannot be acquired.
+
+config MCUMGR_GRP_FS_FILE_AUTOMATIC_IDLE_CLOSE
+	bool "Automatic file handle close on idle"
+	default y
+	help
+	  This feature uses the system workqueue to close fs_mgmt file handles
+	  that have been opened by either a file upload or download command
+	  that has not uploaded or downloaded the full file data (the file
+	  handle is kept open between requests to improve performance and
+	  throughput).
```

Comment:
Have you thought about doing new behaviour default and non-configurable?
What are the benefits of letting this be configurable?

Reply:
OK, so let leave it configurable then.

Reply:
I think this is now not possible to make configurable, made the changes above but the problem is that because you only know the full upload size in the first packet, if you then lose the context, you are not able to know if you have finished uploading the full file size accurately or not (if another request comes in on another transport, you've lost the state). Unless this was made to only allow user configuration if only 1 transport is selected, but I don't think that'd be possible using Kconfig

Reply:
Or:

1. We add a new optional "last" (bool) field to the upload, if present and true, close the file - this will need clients to be changed (would be optional, existing tools would still work)
2. The `FS_MGMT_ID_OPENED_FILE` command is made mandatory in the build and clients, once finished, manually close the opened file using that command (again would be optional, existing tools would still work)

As things are now, if the state is lost, it is not so important because of the lack of file locks, however if we assume we will have file locking in the future, this does become an issue since if MCUmgr does not know when it can close a file, that means that the file close timeout (5 seconds) would need to be exceeded before another transport or the actual application itself, can do anything with that file - this would be a problem.

Reply:
Or we can have those who can not afford extra flash/ram use the old open-do-close, without catching, type of operation.
How much extra flash/ram does the thing need?

Reply:
smp_svr on cortex m4 with overlay-fs, pre:
```
Memory region         Used Size  Region Size  %age Used
           FLASH:       72920 B       896 KB      7.95%
             RAM:       17504 B       256 KB      6.68%
        IDT_LIST:          0 GB         2 KB      0.00%
```
New, with auto close enabled:
```
Memory region         Used Size  Region Size  %age Used
           FLASH:       73924 B       896 KB      8.06%
             RAM:       17656 B       256 KB      6.74%
        IDT_LIST:          0 GB         2 KB      0.00%
```
New, with auto close disabled:
```
Memory region         Used Size  Region Size  %age Used
           FLASH:       73460 B       896 KB      8.01%
             RAM:       17608 B       256 KB      6.72%
        IDT_LIST:          0 GB         2 KB      0.00%
```
Flash: +1KB with auto close, +500 bytes without
RAM: +152 bytes with auto close, +104 bytes without

Reply:
:(.
With such increment we should keep it possible to have it work the old way (where flash gets tortured), for those who can not afford flash/ram and are ok with current lack of performance.
The additional close command is fine.
The problem with lack of auto close, that I have, is that when we get some longer pause in transfer, for example due to some comm problems, the board will be left with opened file for long. Maybe that is nothing, but I am not sure how it will affect underlying file system drivers.

Reply:
I thought nrf52832 was almost failing to build smp_svr with bluetooth but it seems we actually have quite a bit of spare flash on it, default build with bluetooth (using the new system):
```
Memory region         Used Size  Region Size  %age Used
           FLASH:      176336 B       220 KB     78.27%
             RAM:       50232 B        64 KB     76.65%
        IDT_LIST:          0 GB         2 KB      0.00%
```
Slightly more for a production-ready application without logging:
```
Memory region         Used Size  Region Size  %age Used
           FLASH:      164256 B       220 KB     72.91%
             RAM:       50104 B        64 KB     76.45%
        IDT_LIST:          0 GB         2 KB      0.00%
```

Reply:
OK. So IMHO idle close is mandatory with the improved speed fs mgmt, whether close command exists or not.
But this is your call, we can make it mandatory later when we figure out that it breaks fs somehow.